packages/editor/src/store/reducer.js

```diff
@@ -361,63 +428,114 @@ export const editor = flow( [
 		byClientId( state = {}, action ) {
 			switch ( action.type ) {
 				case 'SETUP_EDITOR_STATE':
-					return getFlattenedBlocks( action.blocks );
+					return getFlattenedBlocksWithoutAttributes( action.blocks );
 
 				case 'RECEIVE_BLOCKS':
 					return {
 						...state,
-						...getFlattenedBlocks( action.blocks ),
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
 					};
 
-				case 'UPDATE_BLOCK_ATTRIBUTES':
+				case 'UPDATE_BLOCK':
 					// Ignore updates if block isn't known
 					if ( ! state[ action.clientId ] ) {
 						return state;
 					}
 
-					// Consider as updates only changed values
-					const nextAttributes = reduce( action.attributes, ( result, value, key ) => {
-						if ( value !== result[ key ] ) {
-							result = getMutateSafeObject( state[ action.clientId ].attributes, result );
-							result[ key ] = value;
-						}
-
-						return result;
-					}, state[ action.clientId ].attributes );
-
-					// Skip update if nothing has been changed. The reference will
-					// match the original block if `reduce` had no changed values.
-					if ( nextAttributes === state[ action.clientId ].attributes ) {
+					// Do nothing if only attributes change.
+					const changes = omit( action.updates, 'attributes' );
+					if ( isEmpty( changes ) ) {
 						return state;
 					}
 
-					// Otherwise merge attributes into state
 					return {
 						...state,
 						[ action.clientId ]: {
 							...state[ action.clientId ],
-							attributes: nextAttributes,
+							...changes,
 						},
 					};
 
+				case 'INSERT_BLOCKS':
+					return {
+						...state,
+						...getFlattenedBlocksWithoutAttributes( action.blocks ),
+					};
+
+				case 'REPLACE_BLOCKS':
+					if ( ! action.blocks ) {
```

Comment:
I know this was inherited, but looking at it, it seems quite wrong. We call `replaceBlocks` with an empty array when, for example, removing a paragraph:

https://github.com/WordPress/gutenberg/blob/b5d4e1527af3f15e350da6be4a57963e51208bec/packages/block-library/src/paragraph/edit.js#L116-L117

If I'm reading this right, this behavior would leave lingering state for the removed block?

Reply:
An empty array is truthy, so it seems to me that this should work correctly, unless I'm missing something?

Reply:
> An empty array is truthy, so it seems to me that this should work correctly, unless I'm missing something?

That's a good point. Makes me wonder why the condition exists at all then 🤷‍♂️